### PR TITLE
Use Patch instead of Update for k8s Service client

### PR DESCRIPTION
**Pull request by skmatti**
This replaces of all instances of update calls of k8s Service resource from this controller.

Also, this uses k8s service helper function to patch Service.

/assign @freehan @prameshj

### CODE REVIEW

File: pkg/utils/common/common.go [hunk 1]
```diff
@@ -136,3 +139,21 @@ func patchIngress(ic client.IngressInterface, oldIngress, newIngress *v1beta1.In
...
+	_, err := svchelpers.PatchService(client, svc, newSvc)
+	return err
+}
+
+// PatchServiceLoadBalancerStatus patches the given service's LoadBalancerStatus
+// based on new service's load-balancer status.
+func PatchServiceLoadBalancerStatus(client coreclient.CoreV1Interface, svc *corev1.Service, newStatus corev1.LoadBalancerStatus) error {
+	newSvc := svc.DeepCopy()
+	newSvc.Status.LoadBalancer = newStatus
+	_, err := svchelpers.PatchService(client, svc, newSvc)
```
Comment (prameshj) [inline pkg/utils/common/common.go:157]: Thanks for reusing this!

File: pkg/utils/common/common.go [hunk 2]
```diff
@@ -136,3 +139,21 @@ func patchIngress(ic client.IngressInterface, oldIngress, newIngress *v1beta1.In
 	klog.V(4).Infof("Patch bytes for ingress %s: %s", ingKey, patchBytes)
 	return ic.Patch(context.TODO(), oldIngress.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
 }
+
+// PatchServiceObjectMetadata patches the given service's metadata based on new
+// service metadata.
+func PatchServiceObjectMetadata(client coreclient.CoreV1Interface, svc *corev1.Service, newObjectMetadata metav1.ObjectMeta) error {
```
Comment (bowei) [inline pkg/utils/common/common.go:145]: Should this be in patch.go?

File: pkg/utils/common/common.go [hunk 3]
```diff
@@ -136,3 +139,21 @@ func patchIngress(ic client.IngressInterface, oldIngress, newIngress *v1beta1.In
...
+	newSvc := svc.DeepCopy()
+	newSvc.ObjectMeta = newObjectMetadata
+	_, err := svchelpers.PatchService(client, svc, newSvc)
+	return err
+}
+
+// PatchServiceLoadBalancerStatus patches the given service's LoadBalancerStatus
+// based on new service's load-balancer status.
+func PatchServiceLoadBalancerStatus(client coreclient.CoreV1Interface, svc *corev1.Service, newStatus corev1.LoadBalancerStatus) error {
```
Comment (bowei) [inline pkg/utils/common/common.go:154]: patch.go?

File: pkg/utils/common/common.go (hunk 2 above)
Comment (bowei) [reply pkg/utils/common/common.go:145]: Is there a way to unit test this?

Comment (skmatti) [reply pkg/utils/common/common.go:145]: Can't add it in patch.go because of import cycle. I will add unit tests.

Comment (skmatti) [reply pkg/utils/common/common.go:145]: Maybe I should refactor finalizers into separate package

Comment (skmatti) [reply pkg/utils/common/common.go]: Moved all the patch functions into patch.go and added unit tests

**Comment by bowei** [comment]
Having both utils/common and utils/ doesn't make sense to me.
How about we just collapse common into utils/

**Comment by skmatti** [comment]
> Having both utils/common and utils/ doesn't make sense to me.
> How about we just collapse common into utils/
Its because we will run into an import cycle if not. Both Namer and Finalizer workflows are in utils. I will move those out of utils and fix the import cycle in a followup PR.

**Comment by skmatti** [comment]
> Having both utils/common and utils/ doesn't make sense to me.
> How about we just collapse common into utils/

I have used a different package `patch` for all patch methods. Currently, there are many import cycles if we merge `common` into utils.
Finalizer -> Utils -> Finalizer
Namer -> Utils -> Namer

We created `common` package so that these import cycles are resolved. Maybe we can clean that up later as it exists from longtime now.

**Comment by bowei** [comment]
I think the only thing that really causes the issue is the import of utils/namer from utils/, which is just one file. If we move that out somewhere else, does that resolve the issue/

**Comment by skmatti** [comment]
> I think the only thing that really causes the issue is the import of utils/namer from utils/, which is just one file. If we move that out somewhere else, does that resolve the issue/

This is the line that references namer (https://github.com/kubernetes/ingress-gce/blob/6f08bb6d47ff9325e57f2800ca487133ce3d92df/pkg/utils/serviceport.go#L29).
If I move this to say `serviceport` package, the import cycle would be 
serviceport -> utils/namer -> utils -> serviceport
because of this(https://github.com/kubernetes/ingress-gce/blob/6f08bb6d47ff9325e57f2800ca487133ce3d92df/pkg/utils/utils.go#L434) 

Same with finalizer workflow because of this (https://github.com/kubernetes/ingress-gce/blob/6f08bb6d47ff9325e57f2800ca487133ce3d92df/pkg/utils/utils.go#L465)

**Comment by skmatti** [comment]
@bowei Can you review this again?. Maybe if its better, I can isolate the refactoring into a separate PR so that the change is small.

**Review by freehan — COMMENTED** [review]
can you rebase?

**Review by freehan — APPROVED** [review]
/lgtm
/approve

**Pull request merged**